File: Base/다익스트라/다익스트라.py
from heapq import heappop, heappush
import sys
sys.stdin = open("다익스트라.txt", "r")

input = sys.stdin.readline
INF = sys.maxsize
n, m = map(int, input().split())
start = int(input())
graph = [[] for i in range(n+1)]
distance = [INF]*(n+1)

for _ in range(m):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))


def dijkstra(start):
    heap = []
    distance[start] = 0
    heappush(heap, (0, start))
    while heap:
        dist, now = heappop(heap)
        if distance[now] < dist:
            continue
        for node in graph[now]:
            cost = dist + node[1]
            if cost < distance[node[0]]:
                distance[node[0]] = cost
                heappush(heap, (cost, node[0]))


dijkstra(start)
for i in range(1, n+1):
    if distance[i] == INF:
        print('INF')
    else:
        print(distance[i])

File: 자스베이스/그래프/다익스트라/다익스트라.js
var input = require("fs")
    .readFileSync("./dev/stdin")
    .toString()
    .trim()
    .split("\n");

// const input = require("fs").readFileSync("./1916.txt").toString().split("\r\n");

class MinHeap {
    heap = [0];

    swap(a, b) {
        [this.heap[a], this.heap[b]] = [this.heap[b], this.heap[a]];
    }
    push(v) {
        this.heap.push(v);
        let i = this.heap.length - 1;
        while (i > 1 && this.heap[Math.floor(i / 2)] > this.heap[i]) {
            this.swap(i, Math.floor(i / 2));
            i = Math.floor(i / 2);
        }
    }

    pop() {
        if (this.heap.length <= 1) return 0;
        let x = this.heap[1];
        this.heap[1] = this.heap[this.heap.length - 1];
        this.heap.pop();
        let i = 1;
        while (i * 2 < this.heap.length) {
            let max = this.heap[i * 2];
            let _i = i * 2;
            if (i * 2 + 1 < this.heap.length && max > this.heap[i * 2 + 1]) {
                max = this.heap[i * 2 + 1];
                _i = i * 2 + 1;
            }
            if (this.heap[i] < max) break;
            this.swap(i, _i);
            i = _i;
        }
        return x;
    }
}

const N = parseInt(input[0]);
const M = parseInt(input[1]);

const INF = Infinity;
const dist = Array(N + 1).fill(INF);
const graph = Array.from(Array(N + 1), () => Array());

for (let i = 2; i < M + 2; i++) {
    const [a, b, c] = input[i].split(" ").map(Number);
    graph[a].push([c, b]);
}
const [A, B] = input[M + 2].split(" ").map(Number);

const heap = new MinHeap();
heap.push([0, A]);
dist[A] = 0;
while (heap.heap.length > 1) {
    let [cost, u] = heap.pop();
    if (dist[u] < cost) continue;
    for (let [w, v] of graph[u]) {
        if (dist[v] > dist[u] + w) {
            dist[v] = dist[u] + w;
            heap.push([dist[v], v]);
        }
    }
}
console.log(dist[B]);
